mri/scripts/gridsearch.py

```diff
@@ -0,0 +1,182 @@
+    # get best runs
+    if metric_direction:
+        best_metric = list_metric.max()
+        best_idx = list_metric.argmax()
+    else:
+        best_metric = list_metric.min()
+        best_idx = list_metric.argmin()
+
+    return best_metric, best_idx
+
+
+def launch_grid(linear_params, regularizer_params, reconstructor_params,
+                optimizer_params, compare_metric_details=None, n_jobs=1,
+                verbose=0, **kwargs):
+    """This function launches off reconstruction for a grid specified
+    through use of kwarg dictionaries.
+
+    Dictionary Convention
+    ---------------------
+    These dictionaries each defined to follow the convention:
+    Each dictionary has a key `init_class` that specifies the
+    initialization class for the operator (exception to
+    this is 'optimizer_params'). Later we have key `kwargs` that holds
+    all the input arguments that can be passed as a keyword dictionary.
+    Each value in this keyword dictionary ,ust be a list of all
+    values you want to search in gridsearch.
+
+    This function finds the search space of parameters and
+    sets up right parameters for '_reconstruct_case' function.
+    Please check the example code for more details.
+
+    Parameters
+    ----------
+    linear_params: dict
+        dictionary for linear operator parameters
+    regularizer_params: dict
+        dictionary for regularizer operator parameters
+    reconstructor_params: dict
+        dictionary for reconstructor operator parameters
+    optimizer_params: dict
+        dictionary for optimizer key word arguments
+    compare_metric_details: dict default None
+        dictionary that holds the metric to be compared and metric
+        direction please refer to `gather_result` documentation.
+        if None, all raw_results are returned and best_idx is None
+    n_jobs: int, default 1
+        number of parallel jobs for execution
+    verbose: int default 0
+        Verbosity level
+        0 => No debug prints
+        1 => View best results if present
+
+    **kwargs: keyword arguments that are passes to reconstruction
+        these holds extra arguments that are passed to
+        '_reconstruct_case' function.
+        important arguments are 'ksapce_data' and 'fourier_params'
+    """
+    # Convert non-list elements to list so that we can create
+    # search space
+    all_reformatted_params = []
+    init_classes = []
+    key_names = []
+    for specific_params in [linear_params, regularizer_params,
+                            reconstructor_params, optimizer_params]:
+        for key, value in specific_params['kwargs'].items():
+            if not isinstance(value, list) and \
+                    not isinstance(value, np.ndarray):
+                specific_params['kwargs'][key] = [value]
+        all_reformatted_params.append(
+            list(specific_params['kwargs'].values()))
+        # Obtain Initialization classes
+        if specific_params != optimizer_params:
```

Comment:
have you checked that this works even when `optimizer_params` has a non-list value?

Reply:
In test we use optimizer_params as:

```python
optimizer_params = {
    'kwargs':{
         'optimization_alg': 'fista'
         'num_iterations': 10,
         'metrics': metrics,
    }
}
```

And above works, I dont know why would you be worried aboput this specific case

Reply:
now that it's inplace there is not a problem anymore